src/marshmallow/fields.py

```diff
@@ -1855,6 +1856,73 @@ class IPv6Interface(IPInterface):
     DESERIALIZATION_CLASS = ipaddress.IPv6Interface
 
 
+class Enum(String):
+
+    default_error_messages = {
+        "unknown": "Must be one of: {choices}.",
+    }
+
+    def __init__(
+        self,
+        enum,
+        *args,
+        **kwargs,
+    ):
+        self.enum = enum
+        self.choices = ", ".join(enum.__members__)
+        super().__init__(*args, **kwargs)
+
+    def _serialize(self, value, attr, obj, **kwargs):
+        if value is None:
+            return None
+        return value.name
+
+    def _deserialize(self, value, attr, data, **kwargs):
+        value = super()._deserialize(value, attr, data, **kwargs)
+        try:
+            return getattr(self.enum, value)
+        except AttributeError as exc:
+            raise self.make_error("unknown", choices=self.choices) from exc
+
+
+class TypedEnum:
+    """Base class for typed Enum fields"""
+
+    default_error_messages = {
+        "unknown": "Must be one of: {choices}.",
+    }
+
+    def __init__(
+        self,
+        enum,
+        *args,
+        **kwargs,
+    ):
+        self.enum = enum
+        self.choices = ", ".join([str(m.value) for m in enum])
+        super().__init__(*args, **kwargs)
+
+    def _serialize(self, value, attr, obj, **kwargs):
+        if value is None:
+            return None
+        return value.value
+
+    def _deserialize(self, value, attr, data, **kwargs):
+        value = super()._deserialize(value, attr, data, **kwargs)
+        try:
+            return self.enum(value)
```

Comment:
The main difference between `Enum` and `TypedEnum` seems to be that the former expects enum symbols, while the latter expects enum values. The field names don't currently communicate that.

Reply:
Well `TypedEnum` is not meant to be used, I just didn't make it private to allow for other user-defined typed enums. So it is more about `Enum` vs. `StringEnum` and `IntegerEnum`.

But I agree the names don't really convey the meaning. I'm open to suggestions. Docstrings will help clarifying.

Reply:
I would probably avoid preferring one use case over another with the shorthand `Enum`. `EnumSymbol` and `EnumValue` would be clear regardless of what the user's intuition is.

Reply:
Why not?

So should it be

- `EnumSymbol`
- `EnumValue`
- `EnumString`
- `EnumInteger`

?

Or

- `EnumSymbol`
- `EnumValue`
- `StringEnumValue`
- `IntegerEnumValue`

?

I don't like the former. Kinda like the latter.

Reply:
What about `EnumValue(String)` and `EnumValue(Integer)`? That is consistent with `Nested`/`List` API, improves readability, and minimizes the API surface area.

Reply:
Good idea. I shall investigate this.